WalletWasabi/WabiSabi/Backend/Rounds/Arena.cs

```diff
@@ -254,99 +307,144 @@ private void TimeoutAlices()
+
+			// Begin with Alice locked, to serialize requests concerning a
+			// single coin.
+			using (await alice.AsyncLock.LockAsync(cancellationToken).ConfigureAwait(false))
+			{
+				var otherAlice = AlicesByOutpoint.GetOrAdd(coin.Outpoint, alice);
+
+				if (otherAlice != alice)
 				{
 					throw new WabiSabiProtocolException(WabiSabiProtocolErrorCode.AliceAlreadyRegistered);
 				}
-				return InputRegistrationHandler.RegisterInput(
-					Config,
-					request.RoundId,
-					coin,
-					request.OwnershipProof,
-					request.ZeroAmountCredentialRequests,
-					request.ZeroVsizeCredentialRequests,
-					Rounds);
+
+				using (await AsyncLock.LockAsync(cancellationToken).ConfigureAwait(false))
+				{
+					try
+					{
+						var response = InputRegistrationHandler.RegisterInput(
+							Config,
+							request.RoundId,
+							alice,
+							request.ZeroAmountCredentialRequests,
+							request.ZeroVsizeCredentialRequests);
+
+						// Now that alice is in the round, make it available by id.
+						(AlicesById as IDictionary<uint256, Alice>).Add(alice.Id, alice);
+
+						return response;
+					}
+					catch
+					{
+						AlicesByOutpoint.Remove(coin.Outpoint, out _);
+						throw;
+					}
+				}
 			}
 		}
 
 		public async Task ReadyToSignAsync(ReadyToSignRequestRequest request, CancellationToken cancellationToken)
 		{
-			using (await AsyncLock.LockAsync(cancellationToken).ConfigureAwait(false))
+			if (!RoundsById.TryGetValue(request.RoundId, out var round))
 			{
-				if (Rounds.FirstOrDefault(r => r.Id == request.RoundId) is not Round round)
-				{
-					throw new WabiSabiProtocolException(WabiSabiProtocolErrorCode.RoundNotFound, $"Round ({request.RoundId}) not found.");
-				}
+				throw new WabiSabiProtocolException(WabiSabiProtocolErrorCode.RoundNotFound, $"Round ({request.RoundId}) not found.");
+			}
 
-				if (round.Alices.FirstOrDefault(a => a.Id == request.AliceId) is not Alice alice)
-				{
-					throw new WabiSabiProtocolException(WabiSabiProtocolErrorCode.AliceNotFound, $"Round ({request.RoundId}): Alice id ({request.AliceId}).");
-				}
+			if (!AlicesById.TryGetValue(request.AliceId, out var alice) || alice.Round != round)
+			{
+				throw new WabiSabiProtocolException(WabiSabiProtocolErrorCode.AliceNotFound, $"Round ({request.RoundId}): Alice ({request.AliceId}) not found.");
+			}
 
+			using (await alice.AsyncLock.LockAsync(cancellationToken).ConfigureAwait(false))
+			{
 				var coinJoinInputCommitmentData = new CoinJoinInputCommitmentData("CoinJoinCoordinatorIdentifier", request.RoundId);
 				if (!OwnershipProof.VerifyCoinJoinInputProof(request.OwnershipProof, alice.Coin.TxOut.ScriptPubKey, coinJoinInputCommitmentData))
 				{
 					throw new WabiSabiProtocolException(WabiSabiProtocolErrorCode.WrongOwnershipProof);
 				}
 
-				alice.ReadyToSign = true;
+				using (await AsyncLock.LockAsync(cancellationToken).ConfigureAwait(false))
+				{
+					alice.ReadyToSign = true;
+				}
```

Comment:
It doesn't actually, because boolean types are atomic.


In subsequent, when `ReadyToSign` is triggers `SetPhase` it's necessary to have some sort of protection of the round properties.

Reply:
This part seems easy to get wrong to me and I would say that if the change is needed in the later commits, just add it once it's needed so it can be actually reviewed.